ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java

```diff
@@ -410,44 +412,25 @@ private SourceDescription describe(String name, boolean extended) throws KsqlExc
       ));
     }
 
-    List<PersistentQueryMetadata>
-        queries =
-        ksqlEngine
-            .getPersistentQueries()
-            .values()
-            .stream()
-            .filter(meta ->
-                        ((KsqlStructuredDataOutputNode) meta.getOutputNode())
-                            .getKafkaTopicName()
-                            .equals(dataSource
-                                        .getKsqlTopic()
-                                        .getTopicName()))
-            .collect(Collectors.toList());
     return new SourceDescription(
         dataSource,
         extended,
         dataSource.getKsqlTopic().getKsqlTopicSerDe().getSerDe().name(),
-        getReadQueryIds(queries),
-        getWriteQueryIds(queries),
+        getQueries(q -> q.getSourceNames().contains(dataSource.getName())),
+        getQueries(q -> q.getSinkNames().contains(dataSource.getName())),
         ksqlEngine.getTopicClient()
     );
   }
 
-  private List<String> getReadQueryIds(List<PersistentQueryMetadata> queries) {
-    return queries
+  private List<RunningQuery> getQueries(Predicate<PersistentQueryMetadata> predicate) {
+    return ksqlEngine.getPersistentQueries().values()
```

Comment:
How about we add a method on `ksqlEngine`? something like `getPersistentQueryMetadata` or similar so we aren't exposing the internals so much? i.e., we should be trying to push functionality into the objects themselves as much as possible rather than exposing the internal structure?

Reply:
How about an interface for getting a Collection of all queries and an interface for getting a specific query by ID? It seems gratuitous to push the filtering stuff in. I think it's reasonable to return a Collection and leverage the java.util stuff to warp it into the needed form on the client side.